File: web/src/lib/api/mock.ts
import type {
  ApiClient,
  DiscoverFeeds,
  DiscoveryToken,
  Holder,
  OHLCV,
  Quote,
  QuoteRequest,
  Timeframe,
  Token,
  TokenDetail,
  Trade,
  TradeSide,
  TrendingToken,
  WalletHoldings,
  WalletActivity,
  WalletPositions,
} from "./types";
import { SEED_TOKENS, SeedToken, TRADER_LABELS } from "./seed";

/* ---- deterministic RNG so charts/holders are stable across renders ---- */
function hashStr(s: string): number {
  let h = 2166136261;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  return h >>> 0;
}
function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function toToken(s: SeedToken): Token {
  return {
    address: s.address,
    symbol: s.symbol,
    name: s.name,
    logoURI: null,
    priceUsd: s.price,
    change24h: s.change24h,
    change1h: s.change1h,
    marketCap: s.marketCap,
    liquidity: s.liquidity,
    volume24h: s.volume24h,
  };
}

function findSeed(address: string): SeedToken {
  return (
    SEED_TOKENS.find((t) => t.address === address) ?? SEED_TOKENS[0]
  );
}

/** Build a stable sparkline ending near the current price. */
function sparkline(s: SeedToken, points = 24): number[] {
  const rng = mulberry32(hashStr(s.address + "spark"));
  const drift = s.change24h / 100;
  const out: number[] = [];
  let p = s.price / (1 + drift);
  for (let i = 0; i < points; i++) {
    const noise = (rng() - 0.5) * 0.04;
    p = p * (1 + drift / points + noise);
    out.push(p);
  }
  out[out.length - 1] = s.price;
  return out;
}

const TF_CONFIG: Record<Timeframe, { points: number; stepSec: number }> = {
  "1s": { points: 120, stepSec: 1 },
  "5s": { points: 120, stepSec: 5 },
  "30s": { points: 120, stepSec: 30 },
  "1m": { points: 120, stepSec: 60 },
  "10m": { points: 120, stepSec: 600 },
  "1h": { points: 120, stepSec: 3600 },
  "4h": { points: 120, stepSec: 14400 },
};

function buildOHLCV(s: SeedToken, tf: Timeframe): OHLCV[] {
  const { points, stepSec } = TF_CONFIG[tf];
  const rng = mulberry32(hashStr(s.address + tf));
  const endSec = Math.floor(Date.now() / 1000 / stepSec) * stepSec;
  const vol = stepSec <= 60 ? 0.012 : 0.05;

  // Walk backward from the current price so the latest candle matches it.
  const closes: number[] = new Array(points);
  let price = s.price;
  for (let i = points - 1; i >= 0; i--) {
    closes[i] = price;
    const ret = (rng() - 0.5) * 2 * vol + (s.change24h / 100) * (vol / 6);
    price = price / (1 + ret);
  }

  const out: OHLCV[] = [];
  for (let i = 0; i < points; i++) {
    const close = closes[i];
    const open = i === 0 ? close * (1 + (rng() - 0.5) * vol) : closes[i - 1];
    const hi = Math.max(open, close) * (1 + rng() * vol * 0.8);
    const lo = Math.min(open, close) * (1 - rng() * vol * 0.8);
    out.push({
      time: endSec - (points - 1 - i) * stepSec,
      open,
      high: hi,
      low: lo,
      close,
      volume: (s.volume24h / points) * (0.4 + rng()),
    });
  }
  return out;
}

function buildHolders(s: SeedToken): Holder[] {
  const rng = mulberry32(hashStr(s.address + "holders"));
  const totalSupply = s.marketCap / s.price;
  const count = 28;
  const out: Holder[] = [];
  let remaining = 64; // % of supply distributed among top holders
  for (let i = 0; i < count; i++) {
    const share = i < 3 ? remaining * (0.18 + rng() * 0.1) : remaining * (0.03 + rng() * 0.05);
    const pct = Math.max(0.05, Math.min(share, remaining));
    remaining = Math.max(0, remaining - pct);
    out.push({
      rank: i + 1,
      address: randAddr(rng),
      pct,
      valueUsd: (pct / 100) * s.marketCap,
      tokenAmount: (pct / 100) * totalSupply,
    });
  }
  return out;
}

const B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
function randAddr(rng: () => number): string {
  let out = "";
  for (let i = 0; i < 44; i++) out += B58[Math.floor(rng() * B58.length)];
  return out;
}

function buildTrades(s: SeedToken, n = 40): Trade[] {
  // Uses Date.now() so polling surfaces fresh trades (live feel).
  const rng = mulberry32(hashStr(s.address + Math.floor(Date.now() / 4000)));
  const buyBias = 0.5 + Math.max(-0.25, Math.min(0.25, s.change24h / 200));
  const out: Trade[] = [];
  let t = Date.now();
  for (let i = 0; i < n; i++) {
    const side: TradeSide = rng() < buyBias ? "buy" : "sell";
    const usd = Math.exp(rng() * 6) * 12; // ~$12 .. ~$5k
    const labelled = rng() < 0.4;
    t -= Math.floor(rng() * 14000) + 800;
    out.push({
      id: `${s.address}-${i}-${t}`,
      side,
      trader: randAddr(rng),
      traderLabel: labelled
        ? TRADER_LABELS[Math.floor(rng() * TRADER_LABELS.length)]
        : null,
      amountUsd: usd,
      tokenAmount: usd / s.price,
      priceUsd: s.price * (1 + (rng() - 0.5) * 0.01),
      timestamp: t,
      txHash: randAddr(rng),
    });
  }
  return out;
}

function buildDetail(s: SeedToken): TokenDetail {
  const rng = mulberry32(hashStr(s.address + "detail"));
  const totalSupply = s.marketCap / s.price;
  const holders = buildHolders(s);
  const top10Pct = holders.slice(0, 10).reduce((a, h) => a + h.pct, 0);
  return {
    ...toToken(s),
    fdv: s.marketCap * (1 + rng() * 0.15),
    totalSupply,
    holderCount: Math.floor(2000 + rng() * 60000),
    top10Pct,
    description: `${s.name} ($${s.symbol}) is a community-driven token trading on Solana.`,
    website: null,
    twitter: `https://x.com/${s.symbol.toLowerCase()}`,
    bondingCurve: false,
  };
}

const delay = (ms = 120) => new Promise((r) => setTimeout(r, ms));

export class MockClient implements ApiClient {
  async getBanner(): Promise<Token[]> {
    await delay(80);
    return SEED_TOKENS.map(toToken);
  }

  async getDiscover(): Promise<DiscoverFeeds> {
    await delay(120);
    const trending = await this.getTrending();
    const now = Date.now();
    const NAMES = [
      "Solking Inu", "Solana Pepe", "Wojak Cat", "Moon Doge", "Solis Coin",
      "Based God", "Pump King", "Degen Ape", "Rug Survivor", "Floki Sol",
      "Bonk Jr", "Sigma Frog", "Turbo Cat", "Alpha Wolf",
    ];
    const B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
    const randAddr = (r: () => number) => {
      let s = "";
      for (let i = 0; i < 44; i++) s += B58[Math.floor(r() * B58.length)];
      return s;
    };
    const mk = (i: number, status: "new" | "graduating"): DiscoveryToken => {
      const r = mulberry32(hashStr(NAMES[i % NAMES.length] + status + i));
      const sym = NAMES[i % NAMES.length].split(" ")[0].toUpperCase().slice(0, 6);
      return {
        address: randAddr(r),
        symbol: sym,
        name: NAMES[i % NAMES.length],
        logoURI: null,
        marketCap:
          status === "graduating"
            ? 60000 + r() * 20000
            : 3000 + r() * 30000,
        creator: randAddr(r),
        pool: "pump",
        // new tokens are seconds-to-minutes old; refreshed each poll
        createdAt: now - Math.floor(r() * (status === "new" ? 120 : 1800) * 1000),
        status,
      };
    };
    const fresh = Array.from({ length: 14 }, (_, i) => mk(i, "new")).sort(
      (a, b) => b.createdAt - a.createdAt
    );
    const grad = Array.from({ length: 5 }, (_, i) => mk(i + 20, "graduating")).sort(
      (a, b) => b.createdAt - a.createdAt
    );
    const big = trending.filter((t) => t.marketCap >= 10_000_000);
    return { new: fresh, graduating: grad, trending, big };
  }

  async getTrending(): Promise<TrendingToken[]> {
    await delay(120);
    return [...SEED_TOKENS]
      .sort((a, b) => b.volume24h - a.volume24h)
      .map((s, i) => ({
        ...toToken(s),
        rank: i + 1,
        sparkline: sparkline(s),
      }));
  }

  async getToken(address: string): Promise<TokenDetail> {
    await delay(100);
    return buildDetail(findSeed(address));
  }

  async getOHLCV(address: string, tf: Timeframe): Promise<OHLCV[]> {
    await delay(140);
    return buildOHLCV(findSeed(address), tf);
  }

  async getHolders(address: string): Promise<Holder[]> {
    await delay(120);
    return buildHolders(findSeed(address));
  }

  async getTrades(address: string): Promise<Trade[]> {
    await delay(120);
    return buildTrades(findSeed(address));
  }

  async getHoldings(): Promise<WalletHoldings> {
    await delay(120);
    return { solBalance: 10, tokens: [] };
  }

  async getPositions(): Promise<WalletPositions> {
    await delay(120);
    return { positions: [], realizedSol: 0 };
  }

  async getActivity(): Promise<WalletActivity> {
    await delay(120);
    return { deposited: 0, withdrawn: 0, feesSol: 0, items: [] };
  }

  async getQuote(req: QuoteRequest): Promise<Quote> {
    await delay(160);
    const inSeed = findSeed(req.inputMint);
    const outSeed = findSeed(req.outputMint);
    const inUsd = (req.amount / 1e9) * inSeed.price;
    const outAmount = (inUsd / outSeed.price) * 1e9 * (1 - 0.003);
    return {
      inputMint: req.inputMint,
      outputMint: req.outputMint,
      inAmount: req.amount,
      outAmount: Math.floor(outAmount),
      priceImpactPct: 0.08,
      routeLabel: "Jupiter · best route",
    };
  }
}

File: server/internal/mockdata/mock.go
// Package mockdata implements Provider with deterministic, realistic fake data.
// Used when BIRDEYE_API_KEY is not set, so the backend is runnable with no keys
// and the frontend's live client can be exercised end-to-end.
package mockdata

import (
	"context"
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"

	"solismarket/server/internal/types"
)

type seed struct {
	address   string
	symbol    string
	name      string
	price     float64
	change24h float64
	change1h  float64
	marketCap float64
	liquidity float64
	volume24h float64
}

var seeds = []seed{
	{"So11111111111111111111111111111111111111112", "SOL", "Solana", 168.42, 4.21, 0.62, 81_400_000_000, 240_000_000, 2_100_000_000},
	{"EKpQGSJtjMFqKZ9KQanSqYXRcF8fBopzLHYxdM65zcjm", "WIF", "dogwifhat", 2.31, 12.8, 1.94, 2_300_000_000, 41_000_000, 180_000_000},
	{"DezXAZ8z7PnrnRJjz3wXBoRgixCa6xjnB7YaB1pPB263", "BONK", "Bonk", 0.0000241, -3.42, 0.18, 1_700_000_000, 33_000_000, 95_000_000},
	{"7GCihgDB8fe6KNjn2MYtkzZcRjQy3t9GHdC8uHYmW2hr", "POPCAT", "Popcat", 1.42, 22.4, 3.1, 1_390_000_000, 28_000_000, 120_000_000},
	{"MEW1gQWJ3nEXg2qgERiKu7FAFj79PHvQVREQUzScPP5", "MEW", "cat in a dogs world", 0.0081, 8.9, -0.44, 720_000_000, 19_000_000, 54_000_000},
	{"JUPyiwrYJFskUPiHa7hkeR8VUtAeFoSYbKedZNsDvCN", "JUP", "Jupiter", 0.94, 2.1, 0.31, 1_270_000_000, 36_000_000, 88_000_000},
	{"5z3EqYQo9HiCEs3R84RCDMu2n7anpDMxRhdK8PSWmrRC", "BOME", "BOOK OF MEME", 0.0094, -6.7, -1.2, 660_000_000, 14_000_000, 47_000_000},
	{"WENWENvqqNya429ubCdR81ZmD69brwQaaBYY6p3LCpk", "WEN", "Wen", 0.00013, 14.2, 2.7, 49_000_000, 6_400_000, 12_000_000},
	{"HZ1JovNiVvGrGNiiYvEozEVgZ58xaU3RKwX8eACQBCt3", "PYTH", "Pyth Network", 0.41, 1.4, 0.08, 1_480_000_000, 22_000_000, 31_000_000},
	{"jtojtomepa8beP8AuQc6eXt5FriJwfFMwQx2v2f9mCL", "JTO", "Jito", 2.18, -2.3, 0.22, 540_000_000, 17_000_000, 24_000_000},
	{"SoLixXo3Vj6Jpr8h7uWNb6gV9CzMy3zJ6fy2rqC8gW1k", "GIGA", "Solking", 0.0412, 41.2, 6.4, 410_000_000, 11_000_000, 38_000_000},
	{"TROLLwpHQjP1xQ8jvKpQ4mD8gV2yC5n9aZ7bXr3kU4j", "TROLL", "Troll", 0.062, 18.4, -2.1, 62_000_000, 4_900_000, 9_400_000},
	{"MoonkCat9aZ7bXr3kU4jHQjP1xQ8jvKpQ4mD8gV2yC5n", "MCAT", "Monkey Cat", 0.0228, -11.3, -3.4, 22_000_000, 2_100_000, 5_600_000},
	{"LoLxBmvKpQ4mD8gV2yC5n9aZ7bXr3kU4jHQjP1xQ8jv", "LOL", "LMAO Coin", 0.0000915, 67.8, 9.2, 9_100_000, 1_400_000, 4_200_000},
}

var traderLabels = []string{
	"Roman 尺", "Zrool 尺", "Cupsey", "Esee", "SolLord", "deepfuckingvalue",
	"moonboy.sol", "ansem", "cented", "Pow", "Mr. Frog", "Euris",
}

const b58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

// --- deterministic PRNG (xorshift seeded by FNV hash) ---

func hash(s string) uint32 {
	var h uint32 = 2166136261
	for i := 0; i < len(s); i++ {
		h ^= uint32(s[i])
		h *= 16777619
	}
	return h
}

type prng struct{ s uint32 }

func newPRNG(seed string) *prng {
	h := hash(seed)
	if h == 0 {
		h = 1
	}
	return &prng{s: h}
}

func (p *prng) next() float64 {
	p.s ^= p.s << 13
	p.s ^= p.s >> 17
	p.s ^= p.s << 5
	return float64(p.s) / float64(math.MaxUint32)
}

func (p *prng) addr() string {
	var b strings.Builder
	for i := 0; i < 44; i++ {
		b.WriteByte(b58[int(p.next()*float64(len(b58)))%len(b58)])
	}
	return b.String()
}

func toToken(s seed) types.Token {
	return types.Token{
		Address: s.address, Symbol: s.symbol, Name: s.name, LogoURI: nil,
		PriceUsd: s.price, Change24h: s.change24h, Change1h: s.change1h,
		MarketCap: s.marketCap, Liquidity: s.liquidity, Volume24h: s.volume24h,
	}
}

func find(address string) seed {
	for _, s := range seeds {
		if s.address == address {
			return s
		}
	}
	return seeds[0]
}

// Provider is the mock data source.
type Provider struct{}

func New() *Provider { return &Provider{} }

func (Provider) Banner(context.Context) ([]types.Token, error) {
	out := make([]types.Token, len(seeds))
	for i, s := range seeds {
		out[i] = toToken(s)
	}
	return out, nil
}

func (Provider) Trending(context.Context) ([]types.TrendingToken, error) {
	out := make([]types.TrendingToken, 0, len(seeds))
	// sort by volume desc
	sorted := append([]seed(nil), seeds...)
	for i := 0; i < len(sorted); i++ {
		for j := i + 1; j < len(sorted); j++ {
			if sorted[j].volume24h > sorted[i].volume24h {
				sorted[i], sorted[j] = sorted[j], sorted[i]
			}
		}
	}
	for i, s := range sorted {
		out = append(out, types.TrendingToken{
			Token: toToken(s), Rank: i + 1, Sparkline: sparkline(s),
		})
	}
	return out, nil
}

// Holdings returns an empty demo wallet (mock mode has no chain).
func (Provider) Holdings(_ context.Context, _ string) (*types.WalletHoldings, error) {
	return &types.WalletHoldings{SolBalance: 10, Tokens: []types.TokenBalance{}}, nil
}

// Positions has no chain history in mock mode.
func (Provider) Positions(_ context.Context, _ string) (*types.WalletPositions, error) {
	return &types.WalletPositions{Positions: []types.Position{}}, nil
}

// Activity has no chain history in mock mode.
func (Provider) Activity(_ context.Context, _ string) (*types.WalletActivity, error) {
	return &types.WalletActivity{Items: []types.ActivityItem{}}, nil
}

// Broadcast/TxStatus aren't supported in mock mode (no chain).
func (Provider) Broadcast(_ context.Context, _ string) (string, error) {
	return "", errors.New("swaps are not available in mock mode")
}

func (Provider) TxStatus(_ context.Context, _ string) (string, error) {
	return "", errors.New("swaps are not available in mock mode")
}

func (Provider) RPCProxy(_ context.Context, _ []byte) ([]byte, error) {
	return nil, errors.New("rpc is not available in mock mode")
}

// Big returns the large-cap subset of the mock trending list.
func (p Provider) Big() []types.TrendingToken {
	tr, _ := p.Trending(context.Background())
	big := make([]types.TrendingToken, 0, len(tr))
	for _, t := range tr {
		if t.MarketCap >= 10_000_000 {
			t.Rank = len(big) + 1
			big = append(big, t)
		}
	}
	return big
}

func sparkline(s seed) []float64 {
	r := newPRNG(s.address + "spark")
	drift := s.change24h / 100
	const n = 24
	out := make([]float64, n)
	p := s.price / (1 + drift)
	for i := 0; i < n; i++ {
		p *= 1 + drift/n + (r.next()-0.5)*0.04
		out[i] = p
	}
	out[n-1] = s.price
	return out
}

func (Provider) Token(_ context.Context, address string) (*types.TokenDetail, error) {
	s := find(address)
	r := newPRNG(s.address + "detail")
	holders := buildHolders(s)
	top10 := 0.0
	for i := 0; i < 10 && i < len(holders); i++ {
		top10 += holders[i].Pct
	}
	desc := fmt.Sprintf("%s ($%s) is a community-driven token trading on Solana.", s.name, s.symbol)
	tw := "https://x.com/" + strings.ToLower(s.symbol)
	d := &types.TokenDetail{
		Token:       toToken(s),
		FDV:         s.marketCap * (1 + r.next()*0.15),
		TotalSupply: s.marketCap / s.price,
		HolderCount: int(2000 + r.next()*60000),
		Top10Pct:    top10,
		Description: &desc,
		Website:     nil,
		Twitter:     &tw,
	}
	return d, nil
}

var tfConfig = map[types.Timeframe]struct {
	points  int
	stepSec int64
}{
	types.Tf1s:  {120, 1},
	types.Tf5s:  {120, 5},
	types.Tf30s: {120, 30},
	types.Tf1m:  {120, 60},
	types.Tf10m: {120, 600},
	types.Tf1h:  {120, 3600},
	types.Tf4h:  {120, 14400},
}

func (Provider) OHLCV(_ context.Context, address string, tf types.Timeframe) ([]types.OHLCV, error) {
	s := find(address)
	cfg, ok := tfConfig[tf]
	if !ok {
		cfg = tfConfig[types.Tf1m]
	}
	r := newPRNG(s.address + string(tf))
	endSec := (time.Now().Unix() / cfg.stepSec) * cfg.stepSec
	vol := 0.05
	if tf.BucketSeconds() <= 60 {
		vol = 0.012
	}
	closes := make([]float64, cfg.points)
	price := s.price
	for i := cfg.points - 1; i >= 0; i-- {
		closes[i] = price
		ret := (r.next()-0.5)*2*vol + (s.change24h/100)*(vol/6)
		price = price / (1 + ret)
	}
	out := make([]types.OHLCV, cfg.points)
	for i := 0; i < cfg.points; i++ {
		c := closes[i]
		o := c * (1 + (r.next()-0.5)*vol)
		if i > 0 {
			o = closes[i-1]
		}
		hi := math.Max(o, c) * (1 + r.next()*vol*0.8)
		lo := math.Min(o, c) * (1 - r.next()*vol*0.8)
		out[i] = types.OHLCV{
			Time:   endSec - int64(cfg.points-1-i)*cfg.stepSec,
			Open:   o, High: hi, Low: lo, Close: c,
			Volume: (s.volume24h / float64(cfg.points)) * (0.4 + r.next()),
		}
	}
	return out, nil
}

func buildHolders(s seed) []types.Holder {
	r := newPRNG(s.address + "holders")
	total := s.marketCap / s.price
	const count = 28
	out := make([]types.Holder, 0, count)
	remaining := 64.0
	for i := 0; i < count; i++ {
		var share float64
		if i < 3 {
			share = remaining * (0.18 + r.next()*0.1)
		} else {
			share = remaining * (0.03 + r.next()*0.05)
		}
		pct := math.Max(0.05, math.Min(share, remaining))
		remaining = math.Max(0, remaining-pct)
		out = append(out, types.Holder{
			Rank: i + 1, Address: r.addr(), Pct: pct,
			ValueUsd: pct / 100 * s.marketCap, TokenAmount: pct / 100 * total,
		})
	}
	return out
}

func (Provider) Holders(_ context.Context, address string) ([]types.Holder, error) {
	return buildHolders(find(address)), nil
}

func (Provider) Trades(_ context.Context, address string) ([]types.Trade, error) {
	s := find(address)
	r := newPRNG(s.address + strconv.FormatInt(time.Now().Unix()/4, 10))
	buyBias := 0.5 + math.Max(-0.25, math.Min(0.25, s.change24h/200))
	const n = 40
	out := make([]types.Trade, 0, n)
	t := time.Now().UnixMilli()
	for i := 0; i < n; i++ {
		side := "sell"
		if r.next() < buyBias {
			side = "buy"
		}
		usd := math.Exp(r.next()*6) * 12
		t -= int64(r.next()*14000) + 800
		var label *string
		if r.next() < 0.4 {
			l := traderLabels[int(r.next()*float64(len(traderLabels)))%len(traderLabels)]
			label = &l
		}
		hashStr := r.addr()
		out = append(out, types.Trade{
			ID: fmt.Sprintf("%s-%d-%d", s.address, i, t), Side: side,
			Trader: r.addr(), TraderLabel: label, AmountUsd: usd,
			TokenAmount: usd / s.price, PriceUsd: s.price * (1 + (r.next()-0.5)*0.01),
			Timestamp: t, TxHash: &hashStr,
		})
	}
	return out, nil
}
